tests/devices/unit_tests/test_panda_gridscan.py

```diff
@@ -0,0 +1,303 @@
+            *PASSING_LINE_1,
+            PASSING_CONST,
+            True,
+        ),
+        (
+            *PASSING_LINE_1,
+            *PASSING_LINE_2,
+            PASSING_CONST,
+            *PASSING_LINE_1,
+            FAILING_CONST,
+            False,
+        ),
+        (
+            *PASSING_LINE_1,
+            *PASSING_LINE_2,
+            PASSING_CONST,
+            *FAILING_LINE_1,
+            PASSING_CONST,
+            False,
+        ),
+    ],
+)
+def test_scan_within_limits_3d(
+    x_start,
+    x_steps,
+    x_size,
+    y1_start,
+    y_steps,
+    y_size,
+    z1_start,
+    z2_start,
+    z_steps,
+    z_size,
+    y2_start,
+    expected_in_limits,
+):
+    motor_bundle = create_motor_bundle_with_limits(0.0, 10.0)
+    grid_params = PandAGridScanParams(
+        x_start=x_start,
+        x_steps=x_steps,
+        x_step_size=x_size,
+        y1_start=y1_start,
+        y_steps=y_steps,
+        y_step_size=y_size,
+        z1_start=z1_start,
+        z2_start=z2_start,
+        z_steps=z_steps,
+        z_step_size=z_size,
+        y2_start=y2_start,
+    )
+    assert grid_params.is_valid(motor_bundle.get_xyz_limits()) == expected_in_limits
+
+
+@pytest.fixture
+def grid_scan_params():
+    yield PandAGridScanParams(
+        x_steps=10,
+        y_steps=15,
+        z_steps=20,
+        x_step_size=0.3,
+        y_step_size=0.2,
+        z_step_size=0.1,
+        x_start=0,
+        y1_start=1,
+        y2_start=2,
+        z1_start=3,
+        z2_start=4,
+        run_up_distance_mm=0.05,
+    )
+
+
+@pytest.mark.parametrize(
+    "grid_position",
+    [
+        (np.array([-1, 2, 4])),
+        (np.array([11, 2, 4])),
+        (np.array([1, 17, 4])),
+        (np.array([1, 5, 22])),
+    ],
+)
+def test_given_x_y_z_out_of_range_then_converting_to_motor_coords_raises(
+    grid_scan_params: PandAGridScanParams, grid_position
+):
+    with pytest.raises(IndexError):
+        grid_scan_params.grid_position_to_motor_position(grid_position)
+
+
+def test_given_x_y_z_of_origin_when_get_motor_positions_then_initial_positions_returned(
+    grid_scan_params: PandAGridScanParams,
+):
+    motor_positions = grid_scan_params.grid_position_to_motor_position(
+        np.array([0, 0, 0])
+    )
+    assert np.allclose(motor_positions, np.array([0, 1, 4]))
+
+
+@pytest.mark.parametrize(
+    "grid_position, expected_x, expected_y, expected_z",
+    [
+        (np.array([1, 1, 1]), 0.3, 1.2, 4.1),
+        (np.array([2, 11, 16]), 0.6, 3.2, 5.6),
+        (np.array([6, 5, 5]), 1.8, 2.0, 4.5),
+    ],
+)
+def test_given_various_x_y_z_when_get_motor_positions_then_expected_positions_returned(
+    grid_scan_params: PandAGridScanParams,
+    grid_position,
+    expected_x,
+    expected_y,
+    expected_z,
+):
+    motor_positions = grid_scan_params.grid_position_to_motor_position(grid_position)
+    np.testing.assert_allclose(
+        motor_positions, np.array([expected_x, expected_y, expected_z])
+    )
```

Comment:
Should: These are actually testing things in the common grid scan code, which are presumably tested elsewhere, we don't need to test them again. May be true for other parts of this file too